### PR TITLE
Enables plugin bundling

### PR DESCRIPTION
**Pull request by distantnative**
I have started the path towards plugin bundles, packages, modules – whatever we want to call them. Basically, having the possibility to put tags, fields, snippets, plugins... all together in one plugin folder.

This PR enables plugin modules/bundles so far for:
- [x] Tags
- [x] Snippets
- [x] Templates
- [x] Controllers
- [x] Models
- [x] Languages
- [x] autocss/autojs

To register these tags/fields/... that come with the plugin, there is a simple syntax:

```php
kirby()->modules()->register('snippets',    __DIR__ . DS . 'snippets');
kirby()->modules()->register('templates',   __DIR__ . DS . 'templates');
kirby()->modules()->register('tags',        __DIR__ . DS . 'tags');
kirby()->modules()->register('controllers', __DIR__ . DS . 'controllers');
kirby()->modules()->register('models',      __DIR__ . DS . 'models');
kirby()->modules()->register('languages',   __DIR__ . DS . 'languages');
kirby()->modules()->register('autocss', __DIR__ . DS . 'assets' . DS . 'css' . DS . 'templates');
kirby()->modules()->register('autojs', __DIR__ . DS . 'assets' . DS . 'js' . DS . 'templates');
```

screen shot 2016-02-23 at 21 50 06

The PR also adds an easy helper to set up routes for plugin assets:
```php
kirby()->modules()->assets('assets/plugins/veryUniquePluginName', __DIR__ . DS . 'assets');
kirby()->modules()->assets('WhateverYourRouteBaseShoudLookLike', __DIR__ . DS . 'assets');
a
// and then use, e.g. in site/snippets/footer.php
<?= js('assets/plugins/veryUniquePluginName/plugin.js') ?>
<?= js('WhateverYourRouteBaseShoudLookLike/magic.js') ?>
```

Lots of thanks already to the community in the forum for all the concepts, ideas, remarks...

### CODE REVIEW

**Comment by distantnative** [comment]
Would be great to have some more :eyes: for testing!

File: kirby.php [hunk 1]
```diff
@@ -138,11 +141,14 @@ public function defaults() {
...
-        $root = $kirby->roots()->autocss() . DS . $file;
-        $url  = $kirby->urls()->autocss() . '/' . $file;
+        $file    = $kirby->site()->page()->template() . '.css';
+        $root    = $kirby->roots()->autocss() . DS . $file;
+        $url     = $kirby->urls()->autocss() . '/' . $file;
 
-        if(!file_exists($root)) return false;
+        if(!file_exists($root)) {
+          if(!($url = $kirby->modules()->getAsset('autocss', $file)))
+          return false;
```
Comment (lukasbestle) [inline kirby.php]: I think this should be in the same line as the `if`. Could become a new goto fail bug otherwise. ;)
Alternatively use curly braces.

Same for the JS handler.

File: kirby.php [hunk 2]
```diff
@@ -491,20 +509,23 @@ public function extensions() {
    */
   public function models() {
 
-    if(!is_dir($this->roots()->models())) return false;
+    $roots   = $this->modules()->allRoots('models', $this->roots()->models());
```
Comment (lukasbestle) [inline kirby.php]: The `return false` from the old version got lost here. For compatibility, the method should probably return false if it checked all roots and none existed.

File: kirby.php [hunk 3]
```diff
@@ -491,20 +509,23 @@ public function extensions() {
...
-    foreach($files as $file) {
-      if(f::extension($file) != 'php') continue;
-      $name      = f::name($file);
-      $classname = str_replace(array('.', '-', '_'), '', $name . 'page');
-      $load[$classname] = $root . DS . $file;
+      foreach($files as $file) {
+        if(f::extension($file) != 'php') continue;
+        $name      = f::name($file);
+        $classname = str_replace(array('.', '-', '_'), '', $name . 'page');
+        $load[$classname] = $root . DS . $file;
```
Comment (lukasbestle) [inline kirby.php]: The last registered directory will win here and will override models defined elsewhere. Maybe the directory in `site/models` should be preferred so that users can override the models. I honestly don't know.

File: kirby.php [hunk 4]
```diff
@@ -748,12 +775,12 @@ public function launch() {
...
     date_default_timezone_set($this->options['timezone']);
 
-    // load all extensions
-    $this->extensions();
-
     // load all plugins
     $this->plugins();
 
+    // load all extensions
+    $this->extensions();
```
Comment (lukasbestle) [inline kirby.php]: Why is this change needed? And could it maybe break something else?

File: kirby/modules.php [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+  /**
+   * Magic getter for module arrays
+   */
+  public function __call($method, $arguments) {
+    return isset($this->{$method}) ? $this->{$method} : null;
+  }
+
+  /**
+   * Registers a plugin directory for a module type
```
Comment (lukasbestle) [inline kirby/modules.php]: Method and return value descriptions would be great here. :)

File: kirby/modules.php [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+   */
+  public function register($module, $dir) {
+    if(!isset($this->{$module})) return false;
+
+    // prepare for auto-assets
+    if($module === 'autocss' or $module === 'autojs') {
+      $route = 'assets/modules/' . sha1($dir);
+      $dir   = array($dir, $route);
```
Comment (lukasbestle) [inline kirby/modules.php]: What does that do? Isn't `$dir`/`$root` supposed to be a string value in `assets()`?

File: kirby/modules.php [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+  /**
+   * Creates route for plugin assets
+   */
+   public function assets($route, $root) {
+    if($page = page($route)) return false;
+
+    return kirby()->routes(array(
+      array(
+        'pattern' => trim($route, '/') . '/(:all)',
+        'action'  => function($file) use($route, $root) {
```
Comment (lukasbestle) [inline kirby/modules.php]: `$route` is not used in the action function.

File: kirby/modules.php [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+          $path = $root . DS . $file;
+          if(is_file($path)) {
+            $extension = substr(strchr($path, '.'), 1);
+            return new Response(f::read($path), $extension);
```
Comment (lukasbestle) [inline kirby/modules.php]: Doesn't `f::show($path);` work here?

File: kirby/modules.php [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+
+    if($sort === true) sort($files);
+
+    return array_unique($files);
+  }
+
+  public function allRoots($module, $default = null) {
+    $dirs = $this->{$module}();
+    if(is_null($dirs)) $dirs = array();
+    if($default) $dirs[] = $default;
```
Comment (lukasbestle) [inline kirby/modules.php]: I think the default directory should be the first (`array_unshift()`) to be consistent with `getFile()`.

File: kirby/modules.php [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+  public function getFile($module, $file, $ext, $default = null) {
+
+    if($dirs = $this->{$module}()) {
+      // $ext as array without dots
+      $ext = array_map(function($e) {
+        return ltrim($e, '.');
+      }, (array)$ext);
+
+      // add default location as possible source
+      if(!is_null($default)) array_unshift($dirs, $default);
```
Comment (lukasbestle) [inline kirby/modules.php]: Couldn't this method use `allRoots()` to keep it DRY? This line would then be added to the `allRoots()` method.

File: kirby.php (hunk 3 above)
Comment (distantnative) [inline kirby.php]: Since the default directory `$this->roots()->models()`(aka `site/models`) gets added as last directory in `modules()->allRoots()`, this should already work as you describe it.

File: kirby.php (hunk 4 above)
Comment (distantnative) [inline kirby.php]: The modules feature `tags` as its own modules type atm (as there is also `site/tags`), which is handled by the `extensions()` method. That's why we first have to load the plugins which could register the bundled tags. I thought about if it could break anything, but nothing came to mind - but it's def one of the things why we would need a lot of testing.

File: kirby/modules.php (hunk 8 above)
Comment (distantnative) [inline kirby/modules.php]: Nope, throws an error for the css in my tests:
screen shot 2016-03-04 at 00 25 22

File: kirby/modules.php (hunk 9 above)
Comment (distantnative) [inline kirby/modules.php]: We would need to check if this is actually the case, but my train of thought why they differ: For all cases that use `getFile()` it is important that the default location overrules all others. So it is important to check the default location first and stop if the file is found there. All cases that use `allRoots()` however are not looking for a specific file, but actually load all the files in these locations. So it is important that the default location comes last to overwrite all other registered locations.
So both ways ensure that in the different scenarios the default locations overrule all other locations.

File: kirby/modules.php (hunk 10 above)
Comment (distantnative) [inline kirby/modules.php]: In my opinion not - as `getFile()` and `allRoutes()` essentially need different approaches to the location order as explained in my other comment.

File: kirby.php (hunk 3 above)
Comment (lukasbestle) [inline kirby.php]: Ah, alright. :)
